ipam/ipam.go

```diff
@@ -0,0 +1,252 @@
+package ipam
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	"github.com/giantswarm/e2etests/ipam/provider"
+	"github.com/giantswarm/ipam"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+)
+
+type Config struct {
+	HostFramework *framework.Host
+	Logger        micrologger.Logger
+	Provider      provider.Interface
+
+	CommonDomain    string
+	HostClusterName string
+}
+
+type IPAM struct {
+	hostFramework *framework.Host
+	logger        micrologger.Logger
+	provider      provider.Interface
+
+	commonDomain    string
+	hostClusterName string
+}
+
+func New(config Config) (*IPAM, error) {
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+	if config.CommonDomain == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CommonDomain must not be empty", config)
+	}
+	if config.HostClusterName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostClusterName must not be empty", config)
+	}
+
+	s := &IPAM{
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+		provider:      config.Provider,
+
+		commonDomain:    config.CommonDomain,
+		hostClusterName: config.HostClusterName,
+	}
+
+	return s, nil
+}
+
+func (c *IPAM) Test(ctx context.Context) error {
+
+	var (
+		// Clusters to be created during this test.
+		clusterOne   = c.hostClusterName + "-cluster0"
+		clusterTwo   = c.hostClusterName + "-cluster1"
+		clusterThree = c.hostClusterName + "-cluster2"
+		clusterFour  = c.hostClusterName + "-cluster3"
+
+		// allocatedSubnets[clusterName]subnetCIDRStr
+		allocatedSubnets = make(map[string]string)
+		err              error
+		// guestFrameworks[clusterName]guestFramework
+		guestFrameworks = make(map[string]*framework.Guest)
+	)
+
+	defer func() {
+		c.logger.LogCtx(ctx, "level", "debug", "message", "ensuring all guest clusters possibly created in test are deleted.")
```

Comment:
Nitpicky but I don't think the possibly is needed. For me its still clear without it and easier to read.

Reply:
Check. Removed.